blog/content/post/2022-12-16-trenchboot-aem-for-qubesos.md

```diff
@@ -0,0 +1,367 @@
+
+## Trusted Boot DRTM flow
+
+![Breakout of measured launch details](/img/txt_launch2.jpg)
+
+> Source: *[A Practical Guide to TPM 2.0](https://link.springer.com/book/10.1007/978-1-4302-6584-9)*
+
+The main role of Trusted Boot was to prepare a platform to be launched with
+Intel TXT (Intel's DRTM technology) in a operating system agnostic way. It has
+been achieved by loading a tboot kernel with multiboot protocol and the other
+system components as the modules. That way TBOOT is the main kernel which
+starts first, prepares the platform for TXT launch. When the platform is ready
+then tboot performs the TXT launch. The control is passed to SINIT ACM which
+uses TXT to measure the operating system components in a secure manner. Then
+the control is handed back to tboot kernel which check if the operation was
+successful and boots the target operating system.
+
+Although the tboot tried to be as OS agnostic as possible, some tboot presence
+awareness from operating system is needed because the application processor
+cores (all cores except the main one) are left in a special state after TXT
+launch and cannot be woken up like in traditional boot process. To solve this
+problem, tboot installs a special processor wakeup procedure in the memory
+which OS must call into to start the processor cores. Only then OS may
+initialize the processor per its own requirements.
+
+As one can see the process is complex in case of Intel TXT. Migration of all
+tboot responsibilities was not trivial and has been divided into the work on
+both GRUB and Xen Hypervisor side of Qubes OS.
+
+## GRUB modifications
+
+![](/img/grub_logo.png)
+
+In order to fulfill the same role as tboot, GRUB had to learn how to prepare
+the platform and perform TXT launch. Most of the work for that particular part
+has been done by [Oracle Team working on TrenchBoot for GRUB](https://www.mail-archive.com/grub-devel@gnu.org/msg30167.html).
+That work however covered the Linux kernel TXT launch only. What still had to
+be done was the multiboot protocol support in GRUB to be able to TXT launch a
+Xen Hypervisor. The patches have been prepared for the respective Qubes GRUB
+[package](https://github.com/3mdeb/qubes-grub2/pull/2).
+
+## Xen modifications
+
+![](/img/xen_project_logo.png)
+
+Analogically to GRUB, Xen had to take over some responsibilities from tboot.
+Due to the Intel TXT requirements for boot process, a new entry point had to be
+developed to which SINIT ACM will return control. The new entry point was
+responsible for saving information that a TXT launch happened and clean up the
+processor state so that the booting of Xen kernel could continue with the
+standard multiboot path. Among others, if Xen detected TXT launch, it had to
+perform the special processor cores wakeup process (which has been rewritten
+from TrenchBoot Linux patches to Xen native code) and measure external
+components before using them (that is the Xen parameters, Dom0 Linux kernel and
```

Comment:
and dom0 parameters?

Reply:
Yes, added

Reply:
Resolved as per: https://github.com/3mdeb/news-and-ideas/pull/503#pullrequestreview-1275863118